### DISCUSSION TITLE
What's the state of this project?

### DISCUSSION THREAD

**Discussion by wztvm** [dftworks/dftworks#2] [General]
I'm generally curious what's the status of this project is? You've done some really nice ground work here that could be the basis of a really useful tool. There doesn't seem to have been any commits for a while. Do you have any intention to develop it further and if so would contributions be welcome?

I've interacted with a few pieces of dft software and some of them feel quite dated in their design and it would be cool to have a tool that is just as capable but that makes use of Rust to make the code, configuration, data and interface be a bit more modern and sane.

I'm currently checking out various repos that are related to that endeavour and this one looks promising and is one of the top hits.

Just to get discussion started here are some areas that I think one might look at (just pulled out of the air):
1. Adding and/or improving further introduction and documentation of the project and subprojects. That would include for:
    - the software
    - the config
    - file formats
    - physics
    - maths
    - algorithms
2. Improving configuration.
    - In some cases longer and more descriptive names
    - Maybe compartmentalisation through nesting would be useful.
    - One might use serde to support toml/yaml/json config, and may legacy formats if necessary.
3. Pipeline ergonomics
    - Maybe one could implement something like the Builder-pattern to set it up.
4. Adding/Improving the command line interface
    - Some broad scale config options might more appropriately be cli subcommands.
5. Efficiency (I haven't researched deeply everything the repo currently does so these are just general thoughts)
    - Algorithms
    - Parallelism (cpu, gpu, mpi, clusters)
6. Adding functionality and capability (So that it could be a mature competitor to established tools.) (It may be possible to reuse the ground work here to do a lot of things. That doesn't necessarily mean it all have to go into the same tool, but the more of the ground work that can be reused for similar tasks the better.)
7. Rust macros.
    - A thing I've been thinking constantly when configuring dft runs in other tools is that instead of having every option configurable on the level of the executable it seems like, some of the large scale picture configuration options I might as well be compiled into executables. A workflow that I'd really would like is to be able to spin up a new rust project, import a package, and then in the main function call a builder or macros or some combination there of and maybe wrap it in some manager that generates a cli and file input for the remaining "free" config. The most common workflows might then be made into cli subcommands, or their own commands, or be easily built into any other software like, a desktop application. 
8. Improving the situation for running large numbers of dft calculations either with different config in series or for may different systems in parallell. 
9. Customising data input and output.  
    - supporting different intermediate and output dataformats. (text formats, binary formats, hdf5 etc, databases)
    - reading and writing to and from a database without any temporary disk output.
10. Listeners and events for logging and metrics.

Those are just some brain storming thoughts. Some of it is easy and realistic, other parts may be pipe-dreams. 
If any of it sound interesting, even the simplest stuff, it would be cool to build a community of enthusiasts and make it happen.

Big praise for the work you've done already!!!

**Comment by dftworks** [comment]
Contributions are always welcome! If you are interested, here are several starting points you could consider:

1. Block Davidson eigensolver
2. Exchange-correlation functional: PBE, HSE
3. Projected density of states by angular momentum
4. Spin-orbit coupling: in the spirit of perturbation theory
5. Phonon: finite difference approach
6. GPU support: FFT first, and then eigensolver
7. Interface with Wannier90
8. Interface with BerkeleyGW
9. hdf5 format output for charge density and wavefunction
10. Plotting scripts: band structure, the projected density of states
11. Containerization: docker/singularity image

This project was designed with a focus on functionality, simplicity, performance, and modularity. I intentionally avoided advanced features of Rust and external libraries unless necessary.

**Reply by wztvm** [reply]
Cool, thanks for taking the time to answer.
And great suggestions. I find this topic very interesting so if I can find the time I’ll give one of them a shot, though I might have to read up a bit. If so I’ll probably open a PR for discussion first if you don’t mind.

Nice and well stated focus. I’ll keep those points in mind. =)

**Reply by dftworks** [reply]
Please feel free to open a PR to further discuss a specific feature you feel more comfortable with. Thank you!
